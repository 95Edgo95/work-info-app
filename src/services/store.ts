import createSagaMiddleware, {SagaMiddleware} from "redux-saga";
import {applyMiddleware, compose, createStore} from "redux";
import reducer from "services/reducer";
import saga from "services/saga";
import {fromJS} from "immutable";

let store: any;

const initialState: any = fromJS({});

const sagaMiddleware: SagaMiddleware<any> = createSagaMiddleware();

if (process.env && process.env.NODE_ENV !== "production") {
    const composeEnhancers: any = (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
    store = createStore(reducer, initialState, composeEnhancers(applyMiddleware(sagaMiddleware)));
} else {
    store = createStore(reducer, initialState, compose(applyMiddleware(sagaMiddleware)));
}

sagaMiddleware.run(saga);

export default store;
