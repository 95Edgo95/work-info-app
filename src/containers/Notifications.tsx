import {removeNotification} from "store/notification/NotificationActionCreators";
import {mapDispatchToProps, mapStateToProps} from "helpers/ReduxHelper";
import { ToastContainer, toast } from "react-toastify";
import {ISelectorVariables} from "services/selector";
import {connect} from "react-redux";
import * as React from "react";

import "react-toastify/dist/ReactToastify.css";

interface INotificationsProps {
  removeNotification: (id: number) => void;
  notifications: any;
}

class Notifications extends React.Component<INotificationsProps, undefined> {
  ids: number[] = [];

  componentDidMount(): void {
    const {notifications} = this.props;

    if (notifications.size > 0) {
      this.addNotifications();
    }
  }

  componentDidUpdate(prevProps: any): void {
    const {notifications} = this.props;

    if (!notifications.equals(prevProps.notifications)) {
      this.addNotifications();
    }
  }

  addNotifications() {
    const {notifications} = this.props;

    notifications.forEach(notification => {
      if (this.ids.includes(notification.get("id"))) {
        return;
      }

      switch (notification.get("type")) {
        case "success":
          toast.success(notification.get("message"), {
            onClose: this.onClose(notification.get("id")),
            hideProgressBar: true,
            position: "top-right",
            closeOnClick: true,
            pauseOnHover: true,
          });
          break;

        case "error":
          toast.error(notification.get("message"), {
            onClose: this.onClose(notification.get("id")),
            hideProgressBar: true,
            position: "top-right",
            closeOnClick: true,
            pauseOnHover: true,
          });
          break;

        default:
          toast(notification.get("message"), {
            onClose: this.onClose(notification.get("id")),
            hideProgressBar: true,
            position: "top-right",
            closeOnClick: true,
            pauseOnHover: true,
          });
          break;
      }

      this.ids.push(notification.get("id"));
    });
  }

  onClose = (id: number) => () => {
    const {removeNotification} = this.props;

    const indexOf: number = this.ids.indexOf(id);
    this.ids.splice(indexOf, 1);

    removeNotification(id);
  };

  render(): JSX.Element {
    return (
      <ToastContainer
        hideProgressBar={true}
        position="top-right"
        autoClose={2000}
        pauseOnHover
        closeOnClick
        newestOnTop
      />
    );
  }
}

const selectorVariables: ISelectorVariables = {
  notification: {
    notifications: true,
  }
};

const actions: any = {removeNotification};

export default connect<undefined, INotificationsProps>(mapStateToProps(selectorVariables), mapDispatchToProps(actions))(Notifications);
