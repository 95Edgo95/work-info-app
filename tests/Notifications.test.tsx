import Notifications from "../src/containers/Notifications";
import {BrowserRouter} from "react-router-dom";
import store from "../__mocks__/storeMock";
import {Provider} from "react-redux";
import { mount } from "enzyme";
import * as React from "react";

describe('Layout', () => {
  it('should render correctly', () => {
    const component = mount(
      <Provider store={store}>
        <BrowserRouter>
          <Notifications />
        </BrowserRouter>
      </Provider>
    );

    expect(component).toMatchSnapshot();
    component.unmount();
  });
});
