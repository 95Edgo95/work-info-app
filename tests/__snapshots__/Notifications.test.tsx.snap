// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout should render correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(Notifications)>
        <Notifications
          notifications={Immutable.List []}
          removeNotification={[Function]}
        >
          <ToastContainer
            autoClose={2000}
            bodyClassName={null}
            className={null}
            closeButton={
              <CloseButton
                ariaLabel="close"
              />
            }
            closeOnClick={true}
            draggable={true}
            draggablePercent={80}
            hideProgressBar={true}
            newestOnTop={true}
            pauseOnFocusLoss={true}
            pauseOnHover={true}
            position="top-right"
            progressClassName={null}
            progressStyle={null}
            rtl={false}
            style={null}
            toastClassName={null}
            transition={[Function]}
          >
            <div
              className="Toastify"
            />
          </ToastContainer>
        </Notifications>
      </Connect(Notifications)>
    </Router>
  </BrowserRouter>
</Provider>
`;
